### PR TITLE
Large modular logic refactoring

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

This PR largely rework the logic we use in the modular converter. It is (hopefully) clearer and maintainable. Instead of going in all directions, adding stuff, then deleting it if not needed, we now do the following:

- visit all the modular file (record imports/functions/classes/assignments nodes)
  - create function dependency mapping
- for each import coming from another model:
  - visit the corresponding file
  - create function dependency mapping
  - update mapping with function/assignment from the modular (updated/new functions)
  - create the class dependency graph based on merged dependencies
- update dependency graph of the modular with the functions and assignments imported from the other files
- for each class recorded in the modular:
  - if inherithing from class in another file:
    - replace call to super
    - find the dependencies after the node was replaced
    - follow (updated with modular defs) dependency mapping to add all nodes
  - else:
    - only add needed imported functions (and their dependencies)
- determine the needed imports and add them

Note that we now only visit each files once, instead of potentially revisiting them multiple times due to renaming or deleting nodes at the end.
cc @ArthurZucker for the logic design

Still yet to come if the design looks good:
- Unit-tests!!!

### CODE REVIEW

File: utils/modular_model_converter.py
Comment (ArthurZucker) [inline utils/modular_model_converter.py:976]:
```suggestion
    """This is a Mapper to visit a modular file ( like `modular_llama.py`). It visits the whole file, recording dependency, then visits all imported
    files from the package (`transformers`) and manages their mutual dependencies.
    Calling the method `create_modules()` after visit will create all modules based on this modular file.
    """
```

File: utils/modular_model_converter.py
Comment (ArthurZucker) [inline utils/modular_model_converter.py:994]: not sure what record means in this context

File: utils/modular_model_converter.py
Comment (ArthurZucker) [inline utils/modular_model_converter.py:997]: we need a general variable for this!

Comment (ArthurZucker) [reply utils/modular_model_converter.py:997]: (Like imports_to_skip_in_modular)

File: utils/modular_model_converter.py
Comment (ArthurZucker) [inline utils/modular_model_converter.py:1025]: top-level -> module scoped

File: utils/modular_model_converter.py
Comment (ArthurZucker) [inline utils/modular_model_converter.py:1065]: let's maybe create a separate function for whatever does not need self in this? (would make it more readable)

File: utils/modular_model_converter.py
Comment (ArthurZucker) [inline utils/modular_model_converter.py:1072]:
```suggestion
        1. compute the nested function dependencies
        2. for each model-specific file found in the imports, rename it with the new model name, visit it, and update
        its dependency graph with the new function and assignment definitions found in the modular
        3. update the modular dependency graph with the imported functions and assignments (found when visiting the matching files)
```
would be nice if these are reflected with # 1. , # 2. etc as comments

Comment (ArthurZucker) [reply utils/modular_model_converter.py:1072]: >  for each model-specific file found in the imports, rename it with the new model name: for this we parse the file, and use a FileModelMapper, to visit it and upsate the model name, dependency graph etc

File: utils/modular_model_converter.py
Comment (ArthurZucker) [inline utils/modular_model_converter.py:1128]: does this need to be in this visiter? probably not no?

File: utils/modular_model_converter.py
Comment (ArthurZucker) [inline utils/modular_model_converter.py:1214]: same here, if we create the modules post `leave_Module` I don't think this should be in this class no?

**Review by ArthurZucker — COMMENTED** [review]
Okay! Very nice idea to have a general ModelMapper, and have one for Modular files and one for "other" files. 

Let's keep splitting the functionalities, I think creating the module should go outside the visiter, don't necessarily need a class ! 

Looks good in general! Great work 🔥

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py]: Yes, indeed in all generality it should be a `Name`. I changed and made sure to update at the end so that we don't keep all the garbage Names we don't want (i.e. built-ins, imports etc)

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py:980]: Starting in 3.7, which is end of life now, so should be fine. Anyway, the order is actually not super important here as of now, as the imports are not yet sorted with proper logic (simply relies on the order it is seen in the files, but since we merge imports from different files, the order may end-up being slightly surprising)

**Comment by yonigozlan** [comment]
Hey, so nice to see modular getting better and better! Just a heads up that I have this relatively short PR out on modular as well https://github.com/huggingface/transformers/pull/34477 , which adds some functionalities needed for the ColPali PR, so I was wondering if it will be compatible with this refactoring!
Also made a comment there on adding the modular examples to the check_modular_conversion script to traced when some issues are introduced, I think it could be useful to have here also :)

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py:939]: At some point I was still generating files with duplicated lines of the same imports 
```python
from typing import Optional, Tuple, Union
...
from typing import List
```
which is why I added this quick fix. It does not seem to be needed anymore with latest development, so I removed it for simplicity.
I don't know why it was happening then 😅

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py]: Just removed the dicts to use a list anyway!

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py]: This function was removed

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py:1144]: As the goal anf functionality is the same but the logic is different depending on the type of `Mapper`, I think it makes sense to keep it in the visitors. (It uses a lot of `self` internal variables as well)

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py]: I moved them outside the class and slightly renamed them.

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py]: Yes I was not sure about this choice, I kept it in the class because it was using a lot of internal variables. But as we move more and more stuff outside, let's keep a functional approach.

**Comment by Cyrilvallez** [comment]
Hey @yonigozlan thanks for the heads-up! I added the new `TYPE_TO_FILE_TYPE` and slightly tweaked how I was handling `Annotations`, your `modular_new_kwargs_model.py` now behaves correctly also in this PR.
If you have any issue let me know.

**Review by ArthurZucker — APPROVED** [review]
Very very nice! 
The only things missing now: 
- unittests /  small examples of the capabilities
- update the documentation a little bit to further explain how we do this but mostly 
To be honest we can merge this will unblock me as well 🤗

**Comment by Cyrilvallez** [comment]
Hey @yonigozlan! Thanks for the feedback! I added **assignment dependency tracking** to correctly follow everything (I previously thought it would never be needed, but it turns out it is! And good to have it in general anyway).  
Everything is now behaving nicely, both in the actual models definitions and examples (the examples were broken already btw). Could you double-check just in case? This highlights that we really need those **tests** soon!  
I'll let @ArthurZucker decide if we want to add the examples in the `check_conversion`, but I added them to the `--files_to_parse all` anyway.

**Comment by Cyrilvallez** [comment]
- Added support for a new important use-case (see https://github.com/huggingface/transformers/pull/33770): **correct dispatch of fully new class**. Consider the following `modular.py`:
```
    from ..llama.modeling_llama import LlamaModel

    class NewNameTextConfig(PretrainedConfig):
        ...

    class NewNameConfig(PretrainedConfig):
        ...

    class NewNameModel(LlamaModel):
        config = NewNameConfig()
        text_config = NewNameTextConfig()
        ...
```
we previously had no way of correctly dispatching `NewNameTextConfig` to `configuration_newname.py`, without importing it in `modeling_newname.py` as well as part of the dependencies (because `modeling_llama.py` only tell us that `NewNameConfig` will be imported, but nothing about `TextConfig`).
This is now solved, every fully new class (without exact name match e.g. `LlamaConfig` <-> `NewNameConfig`) which does not belong to the correct file type will not be added, and an import will be created instead.

File: utils/modular_model_converter.py
Comment (Cyrilvallez) [reply utils/modular_model_converter.py:957]: Are we expecting clashes in the names here? Because infering based on the class using them is not so straightforward since different classes in different file types may use them (e.g. type hints)

Comment (Cyrilvallez) [reply utils/modular_model_converter.py:957]: For what I can see, all of them are only used in "processing" files as of now at least

**Comment by ArthurZucker** [comment]
🥳

**Pull request merged**
